:bug: Warn on same layer crossing regardless of bridge / tunnel

homersimpsons opened the pull request:
Fixes #10631

homersimpsons committed:
:bug: Warn on same layer crossing regardless of bridge / tunnel

Fixes #10631

homersimpsons commented:
Shoutout to the project, it was really easy to setup a dev environment on GitHub Codespace and run iD and its test!

Here you can see the new check working:
![image](https://github.com/user-attachments/assets/9849d0db-93fb-4cf1-848e-3ab0c748435e)

- One way is a bridge, the other is not
- Both have the same defined layer
- => The warning is raised

homersimpsons commented on test/spec/validations/crossing_ways.js:
```diff
@@ -373,6 +373,11 @@ describe('iD.validations.crossing_ways', function () {
         verifySingleCrossingIssue(validate(), null);
     });
 
+    it('flags road bridge crossing road on the same layer', function() {
```
I'm not sure about the test positions in this file, but here seems to be the most sensible place.

homersimpsons commented on modules/validations/crossing_ways.js:
```diff
@@ -89,36 +89,21 @@ export function validationCrossingWays(context) {
-            if (layer1 !== layer2) return true;
-        }
-        return false;
+        if (tags1.layer !== undefined && tags1.layer === tags2.layer) return false; // Warn if both have the same defined layer
```
I think I can improve the comments, but I don't really have a great wording at this moment.

tordans commented on modules/validations/crossing_ways.js:
```diff
@@ -89,36 +89,21 @@ export function validationCrossingWays(context) {
+        
+        const isElement1Tunnel = allowsTunnel(featureType1) && hasTag(tags1, 'tunnel');
+        const isElement2Tunnel = allowsTunnel(featureType2) && hasTag(tags2, 'tunnel');
+        if (isElement1Tunnel ^ isElement2Tunnel ) return true; // Either one is tunnel, the other is not
```
Is the ^ correct here? Not ||?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Expressions_and_operators

homersimpsons replied:
Yes, because in case we have 2 `tunnel` crossing we want to warn (in my opinion) if they have the same (or no layer).

I can expend the xor for it to be more readable if you prefer:
```javascript
(isTunnel1 && !isTunnel2) || (!isTunnel1 && isTunnel2) 
```

I will add a test case also specifically for this.

tordans replied:
TIL. Never saw JS [XOR](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_XOR) in the wild. Personally I think using && and || is more common and easier to understand even if a bit longer…

Dimitar5555 replied:
`isElement1Tunnel != isElement2Tunnel` seems clearer and easier to read.

tordans commented on modules/validations/crossing_ways.js:
```diff
@@ -89,36 +89,21 @@ export function validationCrossingWays(context) {
-            // for building crossings, different layers are enough
-            if (layer1 !== layer2) return true;
-        }
-        return false;
```
Is this `building` case really still covered by the code? Are there tests that make sure it is?

homersimpsons replied:
> Is this building case really still covered by the code?

Yes, it falls in both
- the early exit case `if (tags1.layer !== undefined && tags1.layer === tags2.layer) return false;` 
- and the general case `return (tags1.layer || '0') !== (tags2.layer || '0');` 

as those handle anything and hence the `building`s.

> Are there tests that make sure it is?

I believe yes, for example by this test:
https://github.com/openstreetmap/iD/blob/4dc4e7dbdc87a847589c3584f19860bef9ce9aad/test/spec/validations/crossing_ways.js#L151-L155

tordans reviewed (commented):
Just one thing caught my eye. Cannot tell too much else about this without diving deep because I don't know too much about the code and test coverage. This is something for Martin to review IMO.

homersimpsons commented:
Should I improve / change anything here?

> This is something for Martin to review IMO.

Looks like I cannot assign you @tyrasd

homersimpsons commented:
@tyrasd do you want me to do anything to help getting this merged? Like adding tests? Feel free to asks and possibly suggest some changes.

To sum up a bit the changes:
1. The main change is to warn if 2 elements have the same defined `layer` and intersect
2. I rewrote the `tunnel` / `bridge` part to be more concise and explicit (and to remove the case that is already handled by (1))
3. The `building` part is already included in the more general check (1) as discussed in https://github.com/openstreetmap/iD/pull/10999#discussion_r2063755697
4. I added 2 tests for the above changes, all the other tests are untouched

k-yle commented on test/spec/validations/crossing_ways.js:
```diff
@@ -383,6 +388,11 @@ describe('iD.validations.crossing_ways', function () {
+        createWaysWithOneCrossingPoint({ highway: 'residential', tunnel: 'yes', layer: '-1' }, { highway: 'tertiary', layer: '-1' });
+        verifySingleCrossingIssue(validate(), {});
+    });
+
```
```suggestion

    it('flags a crossing where only one feature is indoor', () => {
        createWaysWithOneCrossingPoint({ highway: 'path' }, { highway: 'path', level: '1' });
        verifySingleCrossingIssue(validate(), {});
    });
```
i think one more test case would be good, to confirm that this behaviour is unchanged (1 indoor feature + 1 non-indoor feature)

homersimpsons replied:
thanks for the suggestion, added as-is

k-yle commented on modules/validations/crossing_ways.js:
```diff
@@ -89,36 +89,21 @@ export function validationCrossingWays(context) {
+        const isElement1Bridge = allowsBridge(featureType1) && hasTag(tags1, 'bridge');
+        const isElement2Bridge = allowsBridge(featureType2) && hasTag(tags2, 'bridge');
+        if (isElement1Bridge !== isElement2Bridge) return true; // Either one is bridge, the other is not
+        
```
```suggestion

```
_([from eslint](https://eslint.org/docs/latest/rules/no-trailing-spaces))_

homersimpsons replied:
Thanks (usually my IDE fixes this automatically but looks like GitHub codespace does not no this by default). Fixed and force-pushed.

Note that I didn't delete the entire line as suggested but I just removed the extra spaces as I find that the code is way more readable with the "bridge block" and the "tunnel block". Ping me if you really want to delete the entire line.

k-yle reviewed (approved):
the idea and implementation both look good to me, one suggestion below:

k-yle commented:
if there are no objections, i think this could be merged

k-yle committed:
Merge branch 'develop' into fix/same-layer-crossing-validation

k-yle committed:
add changelog entry

k-yle commented:
I'm going to merge this one, there's been no other comments in the past 6 months, and I think this is an uncontroversial approach to solving the issue, which manages to keep all existing unit tests passing.
